evm_loader/program/src/executor/precompile_extension/metaplex.rs

```diff
@@ -62,60 +62,81 @@ pub fn metaplex<B: AccountStorage>(
                 return Err(Error::StaticModeViolation(*address));
             }
 
-            let mint = read_pubkey(input);
-            let max_supply = read_u64(&input[32..]);
+            let mint = read_pubkey(input)?;
+            let max_supply = read_u64(&input[32..])?;
 
             create_master_edition(context, state, mint, Some(max_supply))
         }
         [0xf7, 0xb6, 0x37, 0xbb] => {
             // "isInitialized(bytes32)"
-            let mint = read_pubkey(input);
+            let mint = read_pubkey(input)?;
             is_initialized(context, state, mint)
         }
         [0x23, 0x5b, 0x2b, 0x94] => {
             // "isNFT(bytes32)"
-            let mint = read_pubkey(input);
+            let mint = read_pubkey(input)?;
             is_nft(context, state, mint)
         }
         [0x9e, 0xd1, 0x9d, 0xdb] => {
             // "uri(bytes32)"
-            let mint = read_pubkey(input);
+            let mint = read_pubkey(input)?;
             uri(context, state, mint)
         }
         [0x69, 0x1f, 0x34, 0x31] => {
             // "name(bytes32)"
-            let mint = read_pubkey(input);
+            let mint = read_pubkey(input)?;
             token_name(context, state, mint)
         }
         [0x6b, 0xaa, 0x03, 0x30] => {
             // "symbol(bytes32)"
-            let mint = read_pubkey(input);
+            let mint = read_pubkey(input)?;
             symbol(context, state, mint)
         }
         _ => Err(Error::UnknownPrecompileMethodSelector(*address, selector)),
     }
 }
 
 #[inline]
-fn read_u64(input: &[u8]) -> u64 {
-    U256::from_be_bytes(*arrayref::array_ref![input, 0, 32]).as_u64()
+fn read_u64(input: &[u8]) -> Result<u64> {
+    if input.len() < 32 {
+        return Err(Error::OutOfBounds);
+    }
+    U256::from_be_bytes(*arrayref::array_ref![input, 0, 32])
+        .try_into()
+        .map_err(Into::into)
 }
 
 #[inline]
-fn read_pubkey(input: &[u8]) -> Pubkey {
-    Pubkey::new_from_array(*arrayref::array_ref![input, 0, 32])
+fn read_pubkey(input: &[u8]) -> Result<Pubkey> {
+    if input.len() < 32 {
+        return Err(Error::OutOfBounds);
+    }
+    Ok(Pubkey::new_from_array(*arrayref::array_ref![input, 0, 32]))
 }
 
 #[inline]
-fn read_string(input: &[u8], offset_position: usize) -> String {
-    let offset = U256::from_be_bytes(*arrayref::array_ref![input, offset_position, 32]).as_usize();
-    let length = U256::from_be_bytes(*arrayref::array_ref![input, offset, 32]).as_usize();
+fn read_string(input: &[u8], offset_position: usize, max_length: usize) -> Result<String> {
+    if input.len() < offset_position + 32 {
+        return Err(Error::OutOfBounds);
+    }
+    let offset =
+        U256::from_be_bytes(*arrayref::array_ref![input, offset_position, 32]).try_into()?;
+    if input.len() < offset + 32 {
+        return Err(Error::OutOfBounds);
```

Comment:
`offset` is controlled by Solidity contract and `offset + 32` can overflow

Reply:
You are right. Fixed